core/market/decentralized/src/protocol/negotiation/provider.rs

```diff
@@ -0,0 +1,238 @@
+            .service(&requestor::proposal_addr(BUS_ID))
+            .send(msg)
+            .await??;
+        Ok(())
+    }
+
+    /// TODO: pass agreement signature.
+    pub async fn approve_agreement(
+        &self,
+        id: NodeId,
+        agreement_id: &str,
+        owner: NodeId,
+    ) -> Result<(), AgreementError> {
+        let msg = AgreementApproved {
+            agreement_id: agreement_id.to_string(),
+        };
+        net::from(id)
+            .to(owner)
+            .service(&requestor::agreement_addr(BUS_ID))
+            .send(msg)
+            .await??;
+        Ok(())
+    }
+
+    pub async fn reject_agreement(
+        &self,
+        id: NodeId,
+        agreement_id: &str,
+        owner: NodeId,
+    ) -> Result<(), AgreementError> {
+        let msg = AgreementRejected {
+            agreement_id: agreement_id.to_string(),
+        };
+        net::from(id)
+            .to(owner)
+            .service(&requestor::agreement_addr(BUS_ID))
+            .send(msg)
+            .await??;
+        Ok(())
+    }
+
+    async fn on_initial_proposal_received(
+        self,
+        caller: String,
+        msg: InitialProposalReceived,
+    ) -> Result<(), ProposalError> {
+        log::debug!(
+            "Negotiation API: Received initial proposal [{}] from [{}].",
+            &msg.proposal_id,
+            &caller
+        );
+        self.inner.initial_proposal_received.call(caller, msg).await
+    }
+
+    async fn on_proposal_received(
+        self,
+        caller: String,
+        msg: ProposalReceived,
+    ) -> Result<(), ProposalError> {
+        log::debug!(
+            "Negotiation API: Received proposal [{}] from [{}].",
+            &msg.proposal_id,
+            &caller
+        );
+        self.inner.proposal_received.call(caller, msg).await
+    }
+
+    async fn on_proposal_rejected(
+        self,
+        caller: String,
+        msg: ProposalRejected,
+    ) -> Result<(), ProposalError> {
+        log::debug!(
+            "Negotiation API: Proposal [{}] rejected by [{}].",
+            &msg.proposal_id,
+            &caller
+        );
+        self.inner.proposal_rejected.call(caller, msg).await
+    }
+
+    async fn on_agreement_received(
+        self,
+        caller: String,
+        msg: AgreementReceived,
+    ) -> Result<(), AgreementError> {
+        log::debug!(
+            "Negotiation API: Agreement proposal [{}] sent by [{}].",
+            &msg.agreement_id,
+            &caller
+        );
+        self.inner.agreement_received.call(caller, msg).await
+    }
+
+    async fn on_agreement_cancelled(
+        self,
+        caller: String,
+        msg: AgreementRejected,
+    ) -> Result<(), AgreementError> {
+        log::debug!(
+            "Negotiation API: Agreement [{}] cancelled by [{}].",
+            &msg.agreement_id,
+            &caller
+        );
+        self.inner.agreement_cancelled.call(caller, msg).await
+    }
+
+    pub async fn bind_gsb(
```

Comment:
pls use `ServiceBinder` for this job

Reply:
What benefit do you see of using it?
Because I can't see any benefit. I will be just forced to get DbExecutor in parameter, that I don't use. But we will have the same amount of code.

Reply:
After discussion I used `ServiceBinder`. Check new version.